File: app/src/main/java/nz/co/codebros/quakesnz/ui/FeatureListActivityModule.java
package nz.co.codebros.quakesnz.ui;

import dagger.Binds;
import dagger.Module;
import dagger.android.ContributesAndroidInjector;
import nz.co.codebros.quakesnz.map.QuakeMapFragment;
import nz.co.codebros.quakesnz.list.QuakeListFragment;
import nz.co.codebros.quakesnz.list.QuakeListModule;
import nz.co.codebros.quakesnz.module.FeatureCollectionModule;
import nz.co.codebros.quakesnz.module.FeatureModule;

/**
 * Created by leandro on 3/07/17.
 */
@Module
public abstract class FeatureListActivityModule {

    @Binds
    abstract QuakeListFragment.OnFeatureClickedListener onFeatureClickedListener(FeatureListActivity activity);

    @ContributesAndroidInjector(modules = {
            QuakeListModule.class,
            FeatureCollectionModule.class,
            FeatureModule.class
    })
    abstract QuakeListFragment quakeListFragment();

    @ContributesAndroidInjector(modules = FeatureModule.class)
    abstract QuakeMapFragment quakeMapFragment();
}

File: app/src/main/java/nz/co/codebros/quakesnz/ui/FeatureListActivityModule.kt
package nz.co.codebros.quakesnz.ui

import android.arch.lifecycle.Transformations
import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import io.reactivex.BackpressureStrategy
import nz.co.codebros.quakesnz.detail.QuakeDetailModel
import nz.co.codebros.quakesnz.list.QuakeListFragment
import nz.co.codebros.quakesnz.map.QuakeMapFragment
import nz.co.codebros.quakesnz.map.QuakeMapState
import nz.co.codebros.quakesnz.scope.FragmentScope
import nz.co.codebros.quakesnz.util.toLiveData

/**
 * Created by leandro on 3/07/17.
 */
@Module
abstract class FeatureListActivityModule {

    @FragmentScope
    @ContributesAndroidInjector
    internal abstract fun quakeListFragment(): QuakeListFragment

    @FragmentScope
    @ContributesAndroidInjector
    internal abstract fun quakeMapFragment(): QuakeMapFragment

    @Module
    companion object {

        @JvmStatic
        @Provides
        fun quakeMapState(
                quakeDetailModel: QuakeDetailModel
        ) = Transformations.map(quakeDetailModel.state, {
            QuakeMapState(it.feature?.geometry?.coordinates)
        })
    }
}
